### PR TITLE
Remove slice_to_array_buffer_view and update_array_buffer_view

### LINKED ISSUE #15350: Rewrite WebVR code to use high level typed array APIs
The WebVR code uses `slice_to_array_buffer_view` and the `ArrayBufferViewContents` trait. It should be rewritten to use the typed array APIs from https://github.com/servo/rust-mozjs/blob/master/src/typedarray.rs instead.

Blocks #14672.

### PR DESCRIPTION
**Pull request by yysung1123**
Remove slice_to_array_buffer_view and update_array_buffer_view

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #15350 
- [X] These changes do not require tests

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15427)

### CODE REVIEW

File: components/script/dom/vrframedata.rs [hunk 1]
```diff
@@ -51,10 +51,14 @@ impl VRFrameData {
...
-            framedata.right_proj.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
-            framedata.right_view.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                MutableHandleObject::from_marked_location(framedata.left_view.get_unsafe()));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                MutableHandleObject::from_marked_location(framedata.left_proj.get_unsafe()));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                MutableHandleObject::from_marked_location(framedata.right_proj.get_unsafe()));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                MutableHandleObject::from_marked_location(framedata.right_view.get_unsafe()));
```
Comment (jdm) [inline components/script/dom/vrframedata.rs]: We should add a `handle_mut()` method to `Heap` instead of using from_marked_location everywhere.

Comment (deror1869107) [inline components/script/dom/vrframedata.rs]: Yes, the code is too long.
I can add ```handle_mut()``` and fix the code in this PR.

**Comment by deror1869107** [comment]
Change all ```MutableHandleObject::from_marked_location``` to ```handle_mut()```.

File: components/script/dom/vreyeparameters.rs [hunk 2]
```diff
@@ -39,7 +39,8 @@ impl VREyeParameters {
         };
 
         unsafe {
-            result.offset.set(slice_to_array_buffer_view(global.get_cx(), &result.parameters.borrow().offset));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), result.parameters.borrow().offset.len() as u32,
+                Some(&result.parameters.borrow().offset), result.offset.handle_mut());
```
Comment (jdm) [inline components/script/dom/vreyeparameters.rs]: Let format this like:
```rust
let _ = Float32Array::create(global.get_cx(),
                             result.parameters.borrow().offset.len() as u32,
                             Some(&result.parameters.borrow().offset),
                             result.offset.handle_mut());
```

File: components/script/dom/vrframedata.rs [hunk 3]
```diff
@@ -51,10 +51,14 @@ impl VRFrameData {
...
-            framedata.right_proj.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
-            framedata.right_view.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                framedata.left_proj.handle_mut());
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                framedata.left_view.handle_mut());
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                framedata.right_proj.handle_mut());
+            let _ = TypedArray::<Float32>::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                framedata.right_view.handle_mut());
```
Comment (jdm) [inline components/script/dom/vrframedata.rs]: nit: use `Float32Array::create` and indent the second line to match `global.get_cx()`.

File: components/script/dom/vrpose.rs [hunk 4]
```diff
@@ -33,9 +33,13 @@ unsafe fn update_or_create_typed_array(cx: *mut JSContext,
     match src {
         Some(ref data) => {
             if dst.get().is_null() {
-                dst.set(slice_to_array_buffer_view(cx, &data));
+                let _ = TypedArray::<Float32>::create(cx, data.len() as u32, src,
+                    dst.handle_mut());
```
Comment (jdm) [inline components/script/dom/vrpose.rs]: nit: `Float32Array::create`, and move `dst.handle_mut()` into the previous line.

File: components/script/dom/vrstageparameters.rs [hunk 5]
```diff
@@ -34,8 +34,10 @@ impl VRStageParameters {
             transform: Heap::default()
         };
         unsafe {
-            stage.transform.set(slice_to_array_buffer_view(global.get_cx(),
-                                                           &stage.parameters.borrow().sitting_to_standing_transform));
+            let _ = TypedArray::<Float32>::create(global.get_cx(),
+                stage.parameters.borrow().sitting_to_standing_transform.len() as u32,
+                Some(&stage.parameters.borrow().sitting_to_standing_transform),
+                stage.transform.handle_mut());
```
Comment (jdm) [inline components/script/dom/vrstageparameters.rs]: nit: `Float32Array::create`, and indent each line to match `global.get_cx()`.

**Review by jdm — CHANGES_REQUESTED** [review]
Looks good! I just have some stylistic comments to make it easier to read.

**Comment by deror1869107** [comment]
Fixed

File: components/script/dom/vrframedata.rs [hunk 6]
```diff
@@ -51,10 +51,14 @@ impl VRFrameData {
...
-            framedata.right_proj.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
-            framedata.right_view.set(slice_to_array_buffer_view(global.get_cx(), &matrix));
+            let _ = Float32Array::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                                         framedata.left_proj.handle_mut());
+            let _ = Float32Array::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                                         framedata.left_view.handle_mut());
+            let _ = Float32Array::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                                         framedata.right_proj.handle_mut());
+            let _ = Float32Array::create(global.get_cx(), matrix.len() as u32, Some(&matrix),
+                                         framedata.right_view.handle_mut());
```
Comment (jdm) [inline components/script/dom/vrframedata.rs]: This is a GC hazard right now - the GC does not know about `framedata` yet, so if any of the `Float32Array::create` calls trigger a GC, the existing typed arrays will be GCed and `framedata` will store invalid pointers. We should call `reflect_dom_object` first before creating the arrays.

**Review by jdm — CHANGES_REQUESTED** [review]
Sorry, one last issue that I noticed.

File: components/script/dom/vrframedata.rs (hunk 6 above)
Comment (deror1869107) [inline components/script/dom/vrframedata.rs]: The code that call ```reflect_dom_object``` has been moved before creating the arrays!

**Pull request merged**